### PR TITLE
[rpc module]: server-side close subscription

### PR DESCRIPTION
**Pull request by niklasad1**
If the subscription request is answered by the server and then the server decides to close the subscription in a background task. Then it's not possible to send a response on some request to cancel that subscription.

Thus, this PR adds functionality to send additional close message when the `SubscriptionSink` when it's dropped.

In the client, the changes are made accordingly to redirect this to the user when the subscription is polled.

### CODE REVIEW

File: tests/tests/integration_tests.rs [hunk 1]
```diff
@@ -33,47 +33,57 @@ use jsonrpsee::{
...
-		let next = add_one.next().await.unwrap();
+		let next = add_one.next().with_default_timeout().await.unwrap().unwrap();
 		assert_eq!(next, i);
 	}
 }
 
 #[tokio::test]
 async fn ws_method_call_works() {
-	let server_addr = websocket_server().await;
+	let server_addr = websocket_server().with_default_timeout().await.unwrap();
```
[the remark this answers is not in the document]
Comment (niklasad1) [reply tests/tests/integration_tests.rs]: sorry, I reverted this will make another PR unrelated this PR.

File: tests/tests/helpers.rs [hunk 2]
```diff
@@ -71,6 +71,16 @@ pub async fn websocket_server_with_subscription() -> SocketAddr {
 			})
 			.unwrap();
 
+		module
+			.register_subscription("subscribe_noop", "unsubscribe_noop", |_, sink, _| {
```
Comment (niklasad1) [inline tests/tests/helpers.rs:75]: This in example to show an edge-case.

If the subscription request is answered by the server and then the server decides to close the subscription. Then it's not possible to send a response on some request to cancel the subscription.

I don't think it's a clean solution this case. AFAIU, `jsonrpc` has it too.

Thus, I suggest an ugly way of handling this; just send a `notif response` with some weird type as result then rely on the assumption that the client will not add this type on the `Subscription<T>` then it should fail the client calls `Subscription::next`

File: utils/src/server/rpc_module.rs [hunk 3]
```diff
@@ -258,21 +259,26 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
...
-			let subscribers = self.subscribers.clone();
 			self.methods.methods.insert(
 				unsubscribe_method_name,
-				Box::new(move |id, params, tx, conn| {
+				Box::new(move |id, params, _, conn_id| {
 					let sub_id = params.one()?;
-					subscribers.lock().remove(&(conn, sub_id));
-					send_response(id, tx, "Unsubscribed");
+					if let Some(sink) = subscribers.lock().remove(&(conn_id, sub_id)) {
+						send_response(id, &sink.0, "Unsubscribe");
```
Comment (niklasad1) [inline utils/src/server/rpc_module.rs:280]: revert

File: tests/tests/integration_tests.rs [hunk 4]
```diff
@@ -240,3 +241,19 @@ async fn ws_unsubscribe_releases_request_slots() {
 	let _: Subscription<JsonValue> =
 		client.subscribe("subscribe_hello", JsonRpcParams::NoParams, "unsubscribe_hello").await.unwrap();
 }
+
+#[tokio::test]
+async fn server_should_be_able_to_close_subscriptions() {
```
Comment (niklasad1) [inline tests/tests/integration_tests.rs:246]: this will fails; I just added this to show a point

File: utils/src/server/rpc_module.rs [hunk 5]
```diff
@@ -343,16 +348,76 @@ impl SubscriptionSink {
 		self.inner_send(msg).map_err(Into::into)
 	}
 
-	fn inner_send(&self, msg: String) -> Result<(), Error> {
-		if self.is_online() {
-			self.inner.unbounded_send(msg).map_err(|e| Error::Internal(e.into_send_error()))
+	fn inner_send(&mut self, msg: String) -> Result<(), Error> {
+		let res = if let Some(online) = self.keep_alive.as_ref() {
+			if online.is_canceled() {
+				return Err(Error::SubscriptionClosed(None));
```
Comment (niklasad1) [inline utils/src/server/rpc_module.rs:354]: this needs to improved.

File: tests/tests/integration_tests.rs [hunk 6]
```diff
@@ -241,3 +242,22 @@ async fn ws_unsubscribe_releases_request_slots() {
...
+	let server_addr = websocket_server_with_subscription().await;
+	let server_url = format!("ws://{}", server_addr);
+
+	let client = WsClientBuilder::default().max_concurrent_requests(1).build(&server_url).await.unwrap();
```
Comment (niklasad1) [inline tests/tests/integration_tests.rs:251]: bad copy-paste

Comment (niklasad1) [inline tests/tests/integration_tests.rs:251]:
```suggestion
	let client = WsClientBuilder::default().build(&server_url).await.unwrap();
```

File: utils/src/server/rpc_module.rs [hunk 7]
```diff
@@ -275,21 +276,25 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
...
-			let subscribers = self.subscribers.clone();
 			self.methods.callbacks.insert(
 				unsubscribe_method_name,
-				MethodCallback::Sync(Box::new(move |id, params, tx, conn| {
+				MethodCallback::Sync(Box::new(move |id, params, tx, conn_id| {
 					let sub_id = params.one()?;
-					subscribers.lock().remove(&(conn, sub_id));
-					send_response(id, tx, "Unsubscribed");
+					subscribers.lock().remove(&(conn_id, sub_id));
+					send_response(id, &tx, "Unsubscribe");
```
Comment (niklasad1) [inline utils/src/server/rpc_module.rs:297]:
```suggestion
					send_response(id, &tx, "Unsubscribed");
```

File: utils/src/server/rpc_module.rs [hunk 8]
```diff
@@ -343,19 +348,85 @@ impl SubscriptionSink {
...
 	}
 
-	fn inner_send(&self, msg: String) -> Result<(), Error> {
-		if self.is_online() {
-			self.inner.unbounded_send(msg).map_err(|e| Error::Internal(e.into_send_error()))
+	fn inner_send(&mut self, msg: String) -> Result<(), Error> {
+		let res = if let Some(online) = self.keep_alive.as_ref() {
+			if online.is_canceled() {
+				return Err(subscription_closed_by_client());
+			}
```
Comment (dvdplm) [inline utils/src/server/rpc_module.rs:355]:
```suggestion
		let res = if let Some(keep_alive) = self.keep_alive.as_ref() {
			if keep_alive.is_canceled() {
				return Err(subscription_closed_by_client());
			}
```

File: utils/src/server/rpc_module.rs [hunk 9]
```diff
@@ -343,19 +348,85 @@ impl SubscriptionSink {
 		self.inner_send(msg).map_err(Into::into)
 	}
 
-	fn inner_send(&self, msg: String) -> Result<(), Error> {
-		if self.is_online() {
-			self.inner.unbounded_send(msg).map_err(|e| Error::Internal(e.into_send_error()))
+	fn inner_send(&mut self, msg: String) -> Result<(), Error> {
+		let res = if let Some(online) = self.keep_alive.as_ref() {
+			if online.is_canceled() {
+				return Err(subscription_closed_by_client());
```
Comment (dvdplm) [inline utils/src/server/rpc_module.rs:354]: This early return here, it means we don't run lines 365-367 to close the `KeepAlive`, why is that?

File: utils/src/server/rpc_module.rs [hunk 10]
```diff
@@ -343,19 +348,83 @@ impl SubscriptionSink {
...
 	}
 
-	fn inner_send(&self, msg: String) -> Result<(), Error> {
-		if self.is_online() {
-			self.inner.unbounded_send(msg).map_err(|e| Error::Internal(e.into_send_error()))
+	fn inner_send(&mut self, msg: String) -> Result<(), Error> {
+		let res = if let Some(keep_alive) = self.keep_alive.as_ref() {
+			if !keep_alive.is_canceled() {
+				// unbounded send only fails if the receiver has been dropped.
+				self.inner.unbounded_send(msg).map_err(|_| subscription_closed_by_client())
```
Comment (dvdplm) [inline utils/src/server/rpc_module.rs:355]: 👍

File: utils/src/server/rpc_module.rs [hunk 11]
```diff
@@ -275,21 +276,25 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
 						inner: method_sink.clone(),
 						method: subscribe_method_name,
 						sub_id,
-						is_online: online_tx,
+						keep_alive: Some(KeepAlive {
```
[the remark this answers is not in the document]
Comment (niklasad1) [reply utils/src/server/rpc_module.rs:279]: I think these is a sign of weird abstraction, I will remove this KeepAlive.

File: utils/src/server/rpc_module.rs [hunk 12]
```diff
@@ -33,7 +33,7 @@ pub type MethodSink = mpsc::UnboundedSender<String>;
 
 type Subscribers = Arc<Mutex<FxHashMap<UniqueSubscriptionEntry, (MethodSink, oneshot::Receiver<()>)>>>;
 
-/// Represent a unique subscription entry based on [`SubscriptionID`] and [`ConnectionID`].
+/// Represent a unique subscription entry based on SubscriptionID and ConnectionID.
```
Comment (dvdplm) [inline utils/src/server/rpc_module.rs:36]: The rustdoc complaint was caused by a misspelled type name I think, `ID` vs `Id`. So think this should work:

```
Represent a unique subscription entry based on [`SubscriptionId`] and [`ConnectionId`].
```

File: utils/src/server/rpc_module.rs [hunk 13]
```diff
@@ -326,9 +330,13 @@ pub struct SubscriptionSink {
 	/// Method.
 	method: &'static str,
 	/// Subscription ID.
-	sub_id: SubscriptionId,
-	/// Whether the subscription is still connected or not.
-	keep_alive: Option<KeepAlive>,
```
Comment (dvdplm) [inline utils/src/server/rpc_module.rs]: Looks cleaner without this for sure. 👍

File: utils/src/server/rpc_module.rs [hunk 14]
```diff
@@ -31,7 +31,14 @@ pub type SubscriptionId = u64;
 /// Sink that is used to send back the result to the server for a specific method.
 pub type MethodSink = mpsc::UnboundedSender<String>;
 
-type Subscribers = Arc<Mutex<FxHashMap<(ConnectionId, SubscriptionId), (MethodSink, oneshot::Receiver<()>)>>>;
+type Subscribers = Arc<Mutex<FxHashMap<UniqueSubscriptionEntry, (MethodSink, oneshot::Receiver<()>)>>>;
+
+/// Represent a unique subscription entry based on [`SubscriptionID`] and [`ConnectionID`].
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+struct UniqueSubscriptionEntry {
```
Comment (dvdplm) [inline utils/src/server/rpc_module.rs:38]: `SubscriptionKey`?

**Review by dvdplm — APPROVED** [review]

**Review by maciejhirsz — APPROVED** [review]

File: utils/src/server/rpc_module.rs (hunk 12 above)
Comment (niklasad1) [reply utils/src/server/rpc_module.rs]: thanks changed it
